### DISCUSSION TITLE
A minimalist, mostly RFC 959 compliant, asyncio FTP server

### DISCUSSION THREAD

**Discussion by DavesCodeMusings** [micropython/micropython#13016] [Show and tell]
# FTPdLite
I've been working on an async FTP server and I thought I'd share. It's mostly for my own education, but it could be useful for some folks.

## What is it?
A minimalist, mostly RFC 959 compliant, asyncio FTP server for MicroPython-based microcontrollers.

## How do I get it?
You can use mpremote mip install github:DavesCodeMusings/ftpdlite

Or you can visit the project site at https://github.com/DavesCodeMusings/ftpdlite/

## For the love of God, man... why?
FTP is one of the earliest internet protocols. It's a security nightmare, but it's a pretty straightforward way to get files from point A to point B. And MicroPython asyncio is something I've been interested in learning more about. So, why not?

>Plus, There's only about nine and a half hours of daylight this time of year. I need something to keep my sanity.

## What else is there?
It's probably got bugs I don't know about, but for the case of a single user uploading and downloading files, it works. Tested clients include: _FileZilla_ (with simultaneous connections set to 1), _curl_, _wget_, and  _tnftp_. (Windoze _ftp.exe_ works too, but not behind a firewall.)

Using asyncio means it does handle multiple sessions, but only one session per client IP address. So still limited in some ways.

There are some extra features built into the SITE command. Together with the _tnftp_ client's extra commands, it makes for a rather sublime remote file management experience. The sample session below demonstrates some of the functionality.

## Sample Session

```
pi@raspberrypi:~ $ man ftp | head -5
TNFTP(1)                  BSD General Commands Manual                 TNFTP(1)

NAME
     tnftp — Internet file transfer program

pi@raspberrypi:~ $ ftp 192.168.10.57
Connected to 192.168.10.57.
220 FTPdLite (MicroPython)
Name (192.168.10.57:pi): Felicia
331 Password required for Felicia.
Password:
230 Login successful.
Remote system type is UNIX.
Using binary mode to transfer files.
ftp> pwd
Remote directory: /
ftp> ls
227 Entering passive mode =192,168,10,57,192,8
150 /
-rw-r--r--  1  root  root         629  Jan  1  2000  boot.py
-rw-r--r--  1  root  root         156  Jan  1  2000  config.py
-rw-r--r--  1  root  root         564  Nov 18 01:06  device_specs.py
drwxr-xr-x  1  root  root           0  Jan  1  2000  lib/
-rw-r--r--  1  root  root         117  Nov 18 01:06  main.py
drwxr-xr-x  1  root  root           0  Jan  1  2000  pub/
226 Directory list sent.
ftp> cd /pub
250 /pub
ftp> ls
227 Entering passive mode =192,168,10,57,192,1
150 /pub
-rw-r--r--  1  root  root     3407402  Nov 18 01:12  Rick_Astley_-_Never_Gonna_Give_You_Up.mp3
-rw-r--r--  1  root  root          21  Nov 18 01:06  test.txt
226 Directory list sent.
ftp> less test.txt
This is a test file.
ftp> site df
211-Filesystem        Size        Used       Avail   Use%
211-flash          6144KiB     3428KiB     2716KiB    56%
211 End.
ftp> site free
211-         Total       Used      Avail
211-Mem:    133KiB      52KiB      80KiB
211 End.
ftp> site gc
211 OK.
ftp> site free
211-         Total       Used      Avail
211-Mem:    133KiB      26KiB     106KiB
211 End.
ftp> rstat
211-FTPdLite (MicroPython)
211-System date: Nov 18 15:43
211-Uptime: 0 days, 01:25
211-Number of users: 1
211-Connected to: ftpdlite
211-Logged in as: Felicia
211-TYPE: L8, FORM: Nonprint; STRUcture: File; transfer MODE: Stream
211 End.
ftp> quit
221 Bye, Felicia.
```

_Edited 19 Nov 2023 to say FileZilla is now working. And multiple user accounts possible. Woohoo!_

**Comment by peterhinch** [comment]
Nice work! If you can get it working with FileZilla we'd have a useful GUI tool for file management on local devices.

**Reply by DavesCodeMusings** [reply]
Thanks @peterhinch

The Filezilla thing is a head scratcher. It logs in, issues a PASV command (I assume for a LIST), but then the connection closes. Not sure why. I can do the same thing using command-line ftp with no problem.

**Reply by robert-hh** [reply]
Did you configure FileZilla to use only a single session?

**Reply by DavesCodeMusings** [reply]
@robert-hh Yes, I did configure single session. I also have my server respond with "421 Too many connections." when it's already got a client connection from a host IP address.

Interestingly, if I set my Filezilla _Transfer Mode > Transfer Settings_ to Active, everything works as expected. But, of course, Active doesn't work well with firewalls.

So it would seem the problem lies with the PASV port negotiation. What I'm seeing in Filezilla logs (see below) is the connection dropping right after the PASV command is issued. But, this never happens with the _tnftp_ command-line client. With _tnftp_ I can log in, list files, upload files, download files... everything that requires a PASV connection works.

Perhaps Filezilla is unhappy with the way I'm replying.

RFC959 suggests this response: 227 Entering Passive Mode. h1,h2,h3,h4,p1,p2

[D.J. Berstein's guide](https://cr.yp.to/ftp.html) suggests: 227 =h1,h2,h3,h4,p1,p2

I landed on a hybrid approach of: Entering passive mode =h1,h2,h3,h4,p1,p2

Perhaps my PASV format is what's causing Filezilla to drop the connection (though tnftp doesn't mind.) The reason for the leading equal sign was given by Berstein as a compatibility measure for old versions of Netscape. Since browsers don't really support FTP anymore, and Netscape is ancient history, I could just drop the equal sign to see if that helps.

Filezilla log with connection being dropped.
```
2023-11-19 08:57:52 17896 1 Status: Logged in
2023-11-19 08:57:52 17896 1 Status: Retrieving directory listing...
2023-11-19 08:57:52 17896 1 Command: PWD
2023-11-19 08:57:52 17896 1 Response: 257 "/"
2023-11-19 08:57:52 17896 1 Command: TYPE I
2023-11-19 08:57:52 17896 1 Response: 200 Always in binary mode.
2023-11-19 08:57:52 17896 1 Command: PASV
2023-11-19 08:57:52 17896 1 Error: Disconnected from server: ECONNABORTED - Connection aborted
2023-11-19 08:57:52 17896 1 Error: Failed to retrieve directory listing
2023-11-19 08:57:52 17896 1 Status: Disconnected from server
```

**Reply by robert-hh** [reply]
The ftp server here: https://github.com/robert-hh/FTP-Server-for-ESP8266-ESP32-and-PYBD sends ` 227 Entering Passive Mode (h1,h2,h3,h4,p1,p2)\r\n`. And it works with filezilla.

**Reply by DavesCodeMusings** [reply]
Thank you! I will give that a try.

**Reply by DavesCodeMusings** [reply]
@robert-hh You, sir, are at the top of my Christmas card list! I've been banging my head on this one for days.

The parentheses are indeed the solution. Not sure which of the FTP RFCs specifies the parentheses, but it satisfies both Filezilla and tnftp, so I'm going with it. 

Now, with my PASV response of: _Entering passive mode (h1,h2,h3,h4,p1,p2)_ Filezilla logged in, gave me a directory listing, and downloaded a file, all without complaint.

Just have to deal with the fact Filezilla doesn't seem to send a QUIT command on disconnect, but rather a Null string??? That's a strange one. One down, one to go, I suppose. Thanks for the help on the PASV thing. Much appreciated.

**Reply by robert-hh** [reply]
Glad it helped. The code I reference uses as well an empty string (besides \r\n) as quit signal, with similar concerns. I added the following comment:
```
            if len(data) <= 0:
                # No data, close
                # This part is NOT CLEAN; there is still a chance that a
                # closing data connection will be signalled as closing
                # command connection
                log_msg(1, "*** No data, assume QUIT")
                close_client(cl)
                return
```

**Reply by DavesCodeMusings** [reply]
Thanks again, @robert-hh 

This was the final piece of the FileZilla compatibility puzzle. I can now log in, list remote directories, download a file, and quit... all in PASV mode from behind a firewall.

**Reply by skyguy** [reply]
If you modify the parameters of your run method like this:

`def run(self, host="0.0.0.0", port=21, loop=None, debug=False):` 

then it will listen on all interfaces automatically, so that it can be started with just:

`server.run()`

**Reply by DavesCodeMusings** [reply]
Thanks @skyguy . I've made the change.

**Reply by DavesCodeMusings** [reply]
And, I had to revert it. I need the host address to send the connection info in the response to a PASV command. With 0.0.0.0 as the host, it doesn't work.
